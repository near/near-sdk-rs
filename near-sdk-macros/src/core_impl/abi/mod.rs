#[cfg(feature = "__abi-embed")]
mod abi_embed;
#[cfg(feature = "__abi-embed")]
pub use abi_embed::embed;

#[cfg(feature = "__abi-generate")]
mod abi_generator;
#[cfg(feature = "__abi-generate")]
pub use abi_generator::generate;
