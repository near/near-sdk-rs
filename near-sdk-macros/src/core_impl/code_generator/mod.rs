mod attr_sig_info;
pub use attr_sig_info::*;

mod impl_item_method_info;
pub use impl_item_method_info::*;

mod item_trait_info;
pub use item_trait_info::*;

mod trait_item_method_info;
pub use trait_item_method_info::*;

mod item_impl_info;
pub use item_impl_info::*;

mod item_struct_info;
pub use item_struct_info::*;
