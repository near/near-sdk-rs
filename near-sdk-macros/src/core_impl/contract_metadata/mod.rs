use darling::{ast::NestedMeta, Error, FromMeta};
use proc_macro2::TokenStream;
use quote::quote;

#[derive(FromMeta)]
struct MacroConfig {
    contract_metadata: Option<ContractMetadata>,
}

#[derive(serde::Serialize, Default, FromMeta)]
struct ContractMetadata {
    version: Option<String>,
    link: Option<String>,
    #[darling(multiple, rename = "standard")]
    standards: Vec<Standard>,
}

#[derive(FromMeta, serde::Serialize)]
struct Standard {
    standard: String,
    version: String,
}

impl ContractMetadata {
    fn populate(mut self) -> Self {
        if self.version.is_none() {
            let version = std::env::var("CARGO_PKG_VERSION").unwrap_or(String::from(""));
            if !version.is_empty() {
                self.version = Some(version);
            }
        }

        if self.link.is_none() {
            let repo = std::env::var("CARGO_PKG_REPOSITORY").unwrap_or(String::from(""));
            if !repo.is_empty() {
                self.link = Some(repo);
            }
        }

        // adding nep330 if it is not present
        if self.standards.is_empty()
            || self.standards.iter().all(|s| s.standard.to_ascii_lowercase() != "nep330")
        {
            self.standards
                .push(Standard { standard: "nep330".to_string(), version: "1.1.0".to_string() });
        }

        self
    }
}

/// Allows for the injection of the contract source metadata infomation into the contract code as
/// a constant.
pub(crate) fn contract_source_metadata_const(attr: proc_macro::TokenStream) -> TokenStream {
    if attr.to_string().is_empty() {
        let metadata = serde_json::to_string(&ContractMetadata::default().populate())
            .expect("ContractMetadata implements Serialize");

        return quote! {
           pub const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
        };
    }

    let attr_args = match NestedMeta::parse_meta_list(attr.into()) {
        Ok(v) => v,
        Err(e) => {
            return Error::from(e).write_errors();
        }
    };

    let args = match MacroConfig::from_list(&attr_args) {
        Ok(v) => v,
        Err(e) => {
            return e.write_errors();
        }
    };

    let metadata = serde_json::to_string(
        &args
            .contract_metadata
            .expect("Attribute input must be present given standard was followed")
            .populate(),
    )
    .expect("ContractMetadata implements Serialize");

    quote! {
        const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
    }
}
