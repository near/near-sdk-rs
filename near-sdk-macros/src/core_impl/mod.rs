#[cfg(feature = "abi")]
pub(crate) mod abi;
mod code_generator;
mod contract_metadata;
mod event;
mod info_extractor;
mod utils;
pub(crate) use code_generator::*;
pub(crate) use contract_metadata::contract_source_metadata_const;
pub(crate) use contract_metadata::ContractMetadata;
pub(crate) use event::{get_event_version, near_events};
pub(crate) use info_extractor::*;
pub(crate) use utils::{get_error_type_from_status, standardized_error_panic_tokens};
