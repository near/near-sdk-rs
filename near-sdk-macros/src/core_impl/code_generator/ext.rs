use crate::core_impl::{serializer, AttrSigInfo};
use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
use quote::{format_ident, quote, ToTokens};
use syn::{parse_quote, Attribute, Generics, Path, Signature};

/// Generates inner ext code for structs and modules. If intended for a struct, generic details
/// for the struct should be passed in through `generic_details` and the `ext` method will be
/// added as an impl to the struct ident.
pub(crate) fn generate_ext_structs(
    ident: &Ident,
    generic_details: Option<&Generics>,
) -> proc_macro2::TokenStream {
    let name = format_ident!("{}Ext", ident);
    let mut ext_code = quote! {
        /// API for calling this contract's functions in a subsequent execution.
        pub fn ext(account_id: near_sdk::AccountId) -> #name {
            #name {
                account_id,
                deposit: 0,
                static_gas: near_sdk::Gas(0),
                gas_weight: near_sdk::GasWeight::default(),
            }
        }
    };
    if let Some(generics) = generic_details {
        // If ext generation is on struct, make ext function associated with struct not module
        ext_code = quote! {
            impl #generics #ident #generics {
                #ext_code
            }
        };
    }

    quote! {
      #[must_use]
      pub struct #name {
          pub(crate) account_id: near_sdk::AccountId,
          pub(crate) deposit: near_sdk::Balance,
          pub(crate) static_gas: near_sdk::Gas,
          pub(crate) gas_weight: near_sdk::GasWeight,
      }

      impl #name {
          pub fn with_attached_deposit(mut self, amount: near_sdk::Balance) -> Self {
              self.deposit = amount;
              self
          }
          pub fn with_static_gas(mut self, static_gas: near_sdk::Gas) -> Self {
              self.static_gas = static_gas;
              self
          }
          pub fn with_unused_gas_weight(mut self, gas_weight: u64) -> Self {
              self.gas_weight = near_sdk::GasWeight(gas_weight);
              self
          }
      }

      #ext_code
    }
}

/// Non-bindgen attributes on contract methods should not be forwarded to the
/// corresponding `_Ext` methods by default. It may lead to compilation errors
/// or unexpected behavior. For a more detailed motivation, see [#959].
///
/// However, some attributes should be forwarded and they are defined here.
///
/// [#959]: https://github.com/near/near-sdk-rs/pull/959
const FN_ATTRIBUTES_TO_FORWARD: [&str; 1] = [
    // Allow some contract methods to be feature gated, for example:
    //
    // ```
    // impl Contract {
    //     #[cfg(integration_tests)]
    //     pub fn test_method(&mut self) { /* ... */ }
    // }
    // ```
    //
    // In that scenario `ContractExt::test_method` should be included only if
    // `integration_tests` is enabled.
    "cfg",
];

/// Returns whether `attribute` should be forwarded to `_Ext` methods, see
/// [`FN_ATTRIBUTES_TO_FORWARD`].
fn is_fn_attribute_to_forward(attribute: &Attribute) -> bool {
    for to_forward in FN_ATTRIBUTES_TO_FORWARD.iter() {
        let to_forward_ident = Ident::new(to_forward, Span::mixed_site());
        let to_forward_path: Path = parse_quote! { #to_forward_ident };
        if to_forward_path == attribute.path {
            return true;
        }
    }
    false
}

/// Generate methods on <StructName>Ext to enable calling each method.
pub(crate) fn generate_ext_function_wrappers<'a>(
    ident: &Ident,
    methods: impl IntoIterator<Item = &'a AttrSigInfo>,
) -> TokenStream2 {
    let ext_ident = format_ident!("{}Ext", ident);
    let mut res = TokenStream2::new();
    for method in methods {
        res.extend(generate_ext_function(method));
    }
    quote! {
        impl #ext_ident {
            #res
        }
    }
}

fn generate_ext_function(attr_signature_info: &AttrSigInfo) -> TokenStream2 {
    let pat_type_list = attr_signature_info.pat_type_list();
    let serialize =
        serializer::generate_serializer(attr_signature_info, &attr_signature_info.input_serializer);

    let AttrSigInfo { non_bindgen_attrs, ident, original_sig, .. } = attr_signature_info;
    let ident_str = ident.to_string();
    let mut new_non_bindgen_attrs = TokenStream2::new();
    for attribute in non_bindgen_attrs.iter() {
        if is_fn_attribute_to_forward(attribute) {
            attribute.to_tokens(&mut new_non_bindgen_attrs);
        }
    }
    let Signature { generics, .. } = original_sig;
    quote! {
        #new_non_bindgen_attrs
        pub fn #ident #generics(self, #pat_type_list) -> near_sdk::Promise {
            let __args = #serialize;
            near_sdk::Promise::new(self.account_id)
            .function_call_weight(
                #ident_str.to_string(),
                __args,
                self.deposit,
                self.static_gas,
                self.gas_weight,
            )
        }
    }
}

#[rustfmt::skip]
#[cfg(test)]
mod tests {
    use crate::core_impl::ImplItemMethodInfo;

    use super::*;
    use quote::quote;
    use syn::{parse_quote, ImplItemMethod, ItemStruct, Type};

    #[test]
    fn ext_gen() {
        let st: ItemStruct = parse_quote! { struct Test { a: u8 } };
        let actual = generate_ext_structs(&st.ident, Some(&st.generics));
        let expected = quote!(
          #[must_use]
          pub struct TestExt {
              pub(crate) account_id: near_sdk::AccountId,
              pub(crate) deposit: near_sdk::Balance,
              pub(crate) static_gas: near_sdk::Gas,
              pub(crate) gas_weight: near_sdk::GasWeight,
          }
          impl TestExt {
              pub fn with_attached_deposit(mut self, amount: near_sdk::Balance) -> Self {
                  self.deposit = amount;
                  self
              }
              pub fn with_static_gas(mut self, static_gas: near_sdk::Gas) -> Self {
                  self.static_gas = static_gas;
                  self
              }
              pub fn with_unused_gas_weight(mut self, gas_weight: u64) -> Self {
                  self.gas_weight = near_sdk::GasWeight(gas_weight);
                  self
              }
          }
          impl Test {
            /// API for calling this contract's functions in a subsequent execution.
            pub fn ext(account_id: near_sdk::AccountId) -> TestExt {
                TestExt {
                    account_id,
                    deposit: 0,
                    static_gas: near_sdk::Gas(0),
                    gas_weight: near_sdk::GasWeight::default(),
                }
            }
          }
        );
        assert_eq!(expected.to_string(), actual.to_string());
    }

    #[test]
    fn module_ext_gen() {
        let ident: Ident = parse_quote! { Test };
        let actual = generate_ext_structs(&ident, None);
        let expected = quote!(
          #[must_use]
          pub struct TestExt {
              pub(crate) account_id: near_sdk::AccountId,
              pub(crate) deposit: near_sdk::Balance,
              pub(crate) static_gas: near_sdk::Gas,
              pub(crate) gas_weight: near_sdk::GasWeight,
          }
          impl TestExt {
              pub fn with_attached_deposit(mut self, amount: near_sdk::Balance) -> Self {
                  self.deposit = amount;
                  self
              }
              pub fn with_static_gas(mut self, static_gas: near_sdk::Gas) -> Self {
                  self.static_gas = static_gas;
                  self
              }
              pub fn with_unused_gas_weight(mut self, gas_weight: u64) -> Self {
                  self.gas_weight = near_sdk::GasWeight(gas_weight);
                  self
              }
          }
          /// API for calling this contract's functions in a subsequent execution.
          pub fn ext(account_id: near_sdk::AccountId) -> TestExt {
              TestExt {
                  account_id,
                  deposit: 0,
                  static_gas: near_sdk::Gas(0),
                  gas_weight: near_sdk::GasWeight::default(),
              }
          }
        );
        assert_eq!(expected.to_string(), actual.to_string());
    }

    /// Verifies that only whitelisted attributes are forwarded to `_Ext`
    /// methods.
    #[test]
    fn ext_fn_non_bindgen_attrs() {
        let impl_type: Type = parse_quote! { Hello };
        let mut method: ImplItemMethod = parse_quote! {
            #[cfg(target_os = "linux")]
            #[inline]
            #[warn(unused)]
            pub fn method(&self) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, impl_type).unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);

        // Note: only whitelisted non-bindgen attributes are forwarded.
        let expected = quote! {
            #[cfg(target_os = "linux")]
            pub fn method (self,) -> near_sdk::Promise {
                let __args = vec![];
                near_sdk::Promise::new(self.account_id).function_call_weight(
                    "method".to_string(),
                    __args,
                    self.deposit,
                    self.static_gas,
                    self.gas_weight,
                )
            }
        };
        assert_eq!(expected.to_string(), actual.to_string());
    }

    #[test]
    fn ext_basic_json() {
        let impl_type: Type = parse_quote! { Hello };
        let mut method: ImplItemMethod = parse_quote! {
            pub fn method(&self, k: &String) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, impl_type).unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);
        let expected = quote!(
            pub fn method(self, k: &String,) -> near_sdk::Promise {
                let __args = {#[derive(near_sdk :: serde :: Serialize)]
                    #[serde(crate = "near_sdk::serde")]
                    struct Input<'nearinput> {
                        k: &'nearinput String,
                    }
                    let __args = Input { k: &k, };
                    near_sdk::serde_json::to_vec(&__args)
                        .expect("Failed to serialize the cross contract args using JSON.")
                };
                near_sdk::Promise::new(self.account_id).function_call_weight(
                    "method".to_string(),
                    __args,
                    self.deposit,
                    self.static_gas,
                    self.gas_weight,
                )
            }
        );
        assert_eq!(expected.to_string(), actual.to_string());
    }

    #[test]
    fn ext_basic_borsh() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemMethod = parse_quote! {
          pub fn borsh_test(&mut self, #[serializer(borsh)] a: String) {}
        };
        let method_info = ImplItemMethodInfo::new(&mut method, impl_type).unwrap();
        let actual = generate_ext_function(&method_info.attr_signature_info);
        let expected = quote!(
          pub fn borsh_test(self, a: String,) -> near_sdk::Promise {
            let __args = {
              #[derive(near_sdk :: borsh :: BorshSerialize)]
              struct Input<'nearinput> {
                  a: &'nearinput String,
              }
              let __args = Input { a: &a, };
              near_sdk::borsh::BorshSerialize::try_to_vec(&__args)
                  .expect("Failed to serialize the cross contract args using Borsh.")
            };
              near_sdk::Promise::new(self.account_id)
                  .function_call_weight(
                      "borsh_test".to_string(),
                      __args,
                      self.deposit,
                      self.static_gas,
                      self.gas_weight,
                  )
          }
        );
        assert_eq!(expected.to_string(), actual.to_string());
    }
}
