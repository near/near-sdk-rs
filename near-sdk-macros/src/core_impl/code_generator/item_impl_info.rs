use crate::core_impl::ext::generate_ext_function_wrappers;
use crate::core_impl::utils;
use crate::core_impl::ReturnKind;
use crate::ItemImplInfo;
use proc_macro2::TokenStream as TokenStream2;
use quote::{format_ident, quote, ToTokens};
use syn::{spanned::Spanned, Ident};

impl ItemImplInfo {
    /// Generate the code that wraps
    pub fn wrapper_code(&self) -> TokenStream2 {
        let mut res = TokenStream2::new();
        let mut checks = quote! {};
        for method in &self.methods {
            res.extend(method.method_wrapper());
            if let ReturnKind::HandlesResultImplicit { .. } =
                method.attr_signature_info.returns.kind
            {
                let error_type = match &method.attr_signature_info.returns.original {
                    syn::ReturnType::Default => quote! { () },
                    syn::ReturnType::Type(_, ty) => {
                        let error_type = utils::extract_error_type(ty);
                        quote! { #error_type }
                    }
                };
                let method_name = &method.attr_signature_info.ident;
                let check_trait_method_name =
                    format_ident!("assert_implements_my_trait{}", method_name);

                checks.extend(quote! {
                    fn #check_trait_method_name() {
                        let _ = near_sdk::check_contract_error_trait as fn(&#error_type);
                    }
                });
            }
        }
        let current_type = &self.ty;
        res.extend(quote! {
            impl #current_type {
                #checks
            }
        });
        res
    }

    pub fn generate_ext_wrapper_code(&self) -> TokenStream2 {
        match syn::parse::<Ident>(self.ty.to_token_stream().into()) {
            Ok(n) => generate_ext_function_wrappers(
                &n,
                self.methods.iter().map(|m| &m.attr_signature_info),
            ),
            Err(e) => syn::Error::new(self.ty.span(), e).to_compile_error(),
        }
    }

    pub fn generate_error_methods(&self) -> TokenStream2 {
        let mut error_methods = quote! {};

        self.methods.iter().map(|m| &m.attr_signature_info).for_each(|method| {
            let error_method_name = quote::format_ident!("{}_error", method.ident);

            if let ReturnKind::HandlesResultImplicit(status) = &method.returns.kind {
                if status.persist_on_error {
                    let error_type = crate::get_error_type_from_status(status);
                    let panic_tokens = crate::standardized_error_panic_tokens();

                    let ty = self.ty.to_token_stream();

                    error_methods.extend(quote! {
                        #[near]
                        impl #ty {
                            pub fn #error_method_name(&self, err: #error_type) {
                                #panic_tokens
                            }
                        }
                    });
                }
            }
        });

        error_methods
    }
}
// Rustfmt removes comas.
#[rustfmt::skip]
#[cfg(test)]
mod tests {
    use syn::{parse_quote, ImplItemFn, Type};
    use crate::core_impl::info_extractor::{ImplItemMethodInfo, ItemImplInfo};
    use crate::core_impl::utils::test_helpers::{local_insta_assert_snapshot, pretty_print_syn_str};


    #[test]
    fn trait_implt() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("fn method(&self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, true, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn no_args_no_return_no_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(&self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn owned_no_args_no_return_no_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }


    #[test]
    fn mut_owned_no_args_no_return() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(mut self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn no_args_no_return_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(&mut self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn arg_no_return_no_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(&self, k: u64) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn args_no_return_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn =
            syn::parse_str("pub fn method(&mut self, k: u64, m: Bar) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn args_return_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn =
            syn::parse_str("pub fn method(&mut self, k: u64, m: Bar) -> Option<u64> { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn args_return_ref() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn =
            syn::parse_str("pub fn method(&self) -> &Option<u64> { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn arg_ref() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method(&self, k: &u64) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn arg_mut_ref() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn =
            syn::parse_str("pub fn method(&self, k: &mut u64) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn callback_args() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[private] pub fn method(&self, #[callback_unwrap] x: &mut u64, y: ::std::string::String, #[callback_unwrap] z: ::std::vec::Vec<u8>) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn callback_args_only() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[private] pub fn method(&self, #[callback_unwrap] x: &mut u64, #[callback_unwrap] y: ::std::string::String) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn callback_args_results() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[private] pub fn method(&self, #[callback_result] x: &mut Result<u64, PromiseError>, #[callback_result] y: Result<::std::string::String, PromiseError>) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn callback_args_vec() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[private] pub fn method(&self, #[callback_vec] x: Vec<String>, y: String) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn simple_init() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init]
            pub fn method(k: &mut u64) -> Self { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn init_no_return() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init]
            pub fn method(k: &mut u64) { }
        };
        let actual = ImplItemMethodInfo::new(&mut method, false, impl_type).map(|_| ()).unwrap_err();
        let expected = "Init function must return the contract state.";
        assert_eq!(expected, actual.to_string());
    }

    #[test]
    fn init_ignore_state() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init(ignore_state)]
            pub fn method(k: &mut u64) -> Self { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn init_payable() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init]
            #[payable]
            pub fn method(k: &mut u64) -> Self { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn args_return_mut_borsh() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[result_serializer(borsh)]
            pub fn method(&mut self, #[serializer(borsh)] k: u64, #[serializer(borsh)]m: Bar) -> Option<u64> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn callback_args_mixed_serialization() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[private] pub fn method(&self, #[callback_unwrap] #[serializer(borsh)] x: &mut u64, #[serializer(borsh)] y: ::std::string::String, #[callback_unwrap] #[serializer(json)] z: ::std::vec::Vec<u8>) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn no_args_no_return_mut_payable() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("#[payable] pub fn method(&mut self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn private_method() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("#[private] pub fn private_method(&mut self) { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn handle_result_json() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[handle_result]
            pub fn method(&self) -> Result::<u64, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }
    
    #[test]
    fn handle_result_mut() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[handle_result]
            pub fn method(&mut self) -> Result<u64, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn handle_result_borsh() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[handle_result]
            #[result_serializer(borsh)]
            pub fn method(&self) -> Result<u64, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn handle_result_init() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init]
            #[handle_result]
            pub fn new() -> Result<Self, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn handle_result_init_ignore_state() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[init(ignore_state)]
            #[handle_result]
            pub fn new() -> Result<Self, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn handle_no_self() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = syn::parse_str("pub fn method() { }").unwrap();
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn result_implicit() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            pub fn method(&self) -> Result<u64, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn persist_on_error() {
        let impl_type: Type = syn::parse_str("Hello").unwrap();
        let mut method: ImplItemFn = parse_quote! {
            #[persist_on_error]
            pub fn method(&mut self) -> Result<u64, &'static str> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.method_wrapper();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

    #[test]
    fn generated_method_error() {
        let mut impl_contract: syn::ItemImpl = parse_quote! {
            impl Contract {
                #[persist_on_error]
                pub fn method(&mut self) -> Result<u64, &'static str> { }
            }
        };
        let impl_contract_info = ItemImplInfo::new(&mut impl_contract).unwrap();
        let actual = impl_contract_info.generate_error_methods();
        local_insta_assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }
}
