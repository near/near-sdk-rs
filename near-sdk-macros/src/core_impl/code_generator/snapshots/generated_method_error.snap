---
source: near-sdk-macros/src/core_impl/code_generator/item_impl_info.rs
assertion_line: 436
expression: pretty_print_syn_str(&actual).unwrap()
---
#[near]
impl Contract {
    pub fn method_error(
        &self,
        err: <Result<u64, &'static str> as near_sdk::__private::ResultTypeExt>::Error,
    ) {
        ::near_sdk::env::panic_err(err.into());
    }
}
